app/src/main/java/com/example/android/weatherapp/fragment/HomeScreenFragment.kt

```diff
@@ -26,16 +28,65 @@ class HomeScreenFragment : BaseFragment() {
         val binding: FragmentHomeScreenBinding = DataBindingUtil.inflate(
             inflater, R.layout.fragment_home_screen, container, false)
         val adapter = HomeScreenAdapter()
-        setOnDataUpdatedObserver(binding, adapter)
+        setOnDataFetchedObserver(binding, adapter)
+        setOnLoadingObserver(binding)
+        setOnErrorObserver(binding)
         return binding.root
     }
 
-    private fun setOnDataUpdatedObserver(binding: FragmentHomeScreenBinding, adapter: HomeScreenAdapter ) {
-        viewModel.eventDataUpdated.observe(viewLifecycleOwner, Observer {
+    private fun setOnDataFetchedObserver(binding: FragmentHomeScreenBinding, adapter: HomeScreenAdapter ) {
+        viewModel.eventDataFetched.observe(viewLifecycleOwner, Observer {
             binding.homeScreenReyclerview.adapter = adapter
-            adapter.currentWeatherData = viewModel.currentWeather
-            adapter.threeHourlyWeatherData = viewModel.threeHourlyWeatherForecast
-            adapter.weeklyWeatherData = viewModel.weeklyWeatherForecast
+            adapter.data = viewModel.forecastData
         })
     }
+
+    private fun setOnLoadingObserver(binding: FragmentHomeScreenBinding) {
+        viewModel.eventLoading.observe(viewLifecycleOwner, Observer {hasLoaded ->
+            if(hasLoaded) {
```

Comment:
Make a separate method in base fragment to handle this if-else logic, so that your observer method remains clean.